Add initial support for customizing the catalog import page

Fox32 opened the pull request:
This makes the catalog import easier customizable, similar how it's done in the search page. Till now, one has to create a custom plugin to do this. Otherwise this keeps working as before, there are no visual changes.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

freben commented on plugins/catalog-import/src/components/types.ts:
```diff
@@ -0,0 +1,30 @@
+};
+
+export type StepperApis = {
+  configApi: ConfigApi;
```
(I see this is not your code, so just putting this here as a side thought, no action needed)

Initial reaction when seeing this: do they really need a config API, or do they need a `Config`?

I know that one type is aliased to the other, but in this case as a reader I kinda expected this to be more on the form `type StepperOptions = { config: Config }` or so, unless the stepper _itself_ also was an API that depended on other APIs.

freben reviewed (commented):
Looks OK I think. Not sure about the context - what's the main idea behind using a context here? Would it make sense to just have a `useImportClient` hook or whatever to supply the functionality for making the pull request or whatnot? It could have the config flag too right? I'm probably missing something

Fox32 commented:
The context is required to provide the current configuration. I agree that this isn't nice.

There are two configuration options:
1. Whether it is supported to automatically create a PR that adds a `catalog-info.yaml` to the repository.
2. A callback to customize the PR.

The first option is evaluated in both the stepper component and the info card. For sure this could also be a configuration option accessed via the configuration API. The question is where to draw the line - is that really something that one want to configure, or is it a customization? Another option could be passing the prop to both the stepper and the info card manually, and requiring to create a custom page (now supported with this PR) to disable it.

The second option is only required for the stepper component. Actually, I think that this callback could also be a nice API. It would still be overridable to customize the behavior. One could also argue that one has to create a custom page to set it.

Now that customization is possible, we could move setting both option there. It would feel "more difficult", but I think that would be a nice pattern. But I tried to avoid a breaking change, if we are ok with it, I can refactor it a bit.

Fox32 commented:
I'm tempted to move both into the `CatalogImportClient`. A method like `canCreatePullRequest(): boolean` and a `preparePullRequest(): Promise<…>`. We could also go with an optional `preparePullRequest?(): Promise<…>`, but I think that would be harder to override, especially if the default is non `undefined`. That way we could remove the configuration and let users override the API in their app. No need for the `ImportOptionsContext` then.

freben commented on plugins/catalog-import/api-report.md:
```diff
@@ -146,6 +147,11 @@ export function defaultGenerateStepper(
+// Warning: (ae-missing-release-tag) "DefaultImportComponentPage" is exported by the package, but it is missing a release tag (@alpha, @beta, @public, or @internal)
+//
+// @public (undocumented)
+export const DefaultImportComponentPage: () => JSX.Element;
```
it's unfortunate to use the word "component" here since it's just one of several possible kinds

Fox32 replied:
Agree, I will rename it. Even thought everything else is still referring to components. Like "Register an existing component" or "Start tracking your component in …" 😆

freben commented on plugins/catalog-import/src/components/ImportStepper/ImportStepper.tsx:
```diff
@@ -42,18 +43,19 @@ type Props = {
   const configApi = useApi(configApiRef);
   const classes = useStyles();
   const state = useImportState({ initialUrl });
+  const opts = useImportOptions();
```
if we deprecate rather than remove the `opts` prop, should this code do `const opts = props.opts || useImportOptions()` instead? Ok, but not using the `||` since it's a hook

freben commented on plugins/catalog-import/src/components/ImportInfoCard/ImportInfoCard.tsx:
```diff
@@ -0,0 +1,80 @@
+  const opts = useImportOptions();
+
+  const integrations = configApi.getConfig('integrations');
+  const hasGithubIntegration = integrations.has('github');
```
interestingly (just noting, no action) I think that this can never be false (it inserts a `github.com` one transparently)

Fox32 replied:
I think this works as expected as it directly reads from the configuration file, not the integrations API.

freben commented on plugins/catalog-import/src/components/ImportInfoCard/ImportInfoCard.tsx:
```diff
@@ -0,0 +1,80 @@
+            repository, previews the entities, and adds them to the {appTitle}{' '}
+            catalog.
+          </Typography>
+          {!opts?.pullRequest?.disable && (
```
should this be like `opts?.pullRequest?.preparePullRequest && !opts?.pullRequest?.disable`?

freben reviewed (commented):
Alright, had another look.

I think you may be onto something with the idea of the client. Or, rather...

Where I want to go with the import flow, is instead that it's broken apart into several steps. I'd like it to ask upfront, what do you want to do?

- Create something from a template?
- Register an existing yaml file?
- Import a directory

This page could have the smarts to not show an option that we know isn't supported, based on what integrations exist etc.

The help text on the second page would be tailored to the task chosen.

The import-a-directory wizard would be able to analyze the url and respond to you that hey, sorry but we are currently not able to create pull requests for this version control provider, etc. Then it's not really a context we need, but something that's more on the form of `const prs: boolean = buildActionClassForUrl(url).supportsPullRequests`

Fox32 commented on plugins/catalog-import/api-report.md:
```diff
@@ -205,11 +224,6 @@ export const PreviewPullRequestComponent: ({
-// Warning: (ae-missing-release-tag) "Router" is exported by the package, but it is missing a release tag (@alpha, @beta, @public, or @internal)
-//
-// @public (undocumented)
-export const Router: (opts: StepperProviderOpts) => JSX.Element;
```
It's already deprecated for some time, I think we can remove it.

Fox32 commented on plugins/catalog-import/api-report.md:
```diff
@@ -116,11 +127,10 @@ export class CatalogImportClient implements CatalogImportApi {
 //
 // @public (undocumented)
-export const CatalogImportPage: (opts: StepperProviderOpts) => JSX.Element;
+export const CatalogImportPage: () => JSX.Element;
```
Here is a breaking change. While I could keep the parameters as is, it couldn't find I way to make it work. If I just keep the parameter to keep compatibility, but not keep the logic it would still be a breaking change. So I think going with a breaking change is a better choice.

Fox32 commented:
Updated the PR to use an optional method in the API for configuration.

Fox32 commented:
@freben resolved the conflicts. Anything that we can do here?

freben commented:
I'm off - deferring to @benjdlambert

benjdlambert commented:
I think this looks pretty good are you happy @freben?

freben commented:
I put it on the queue, will give it a look soon

freben commented on plugins/catalog-import/src/api/CatalogImportClient.ts:
```diff
@@ -114,6 +118,24 @@ export class CatalogImportClient implements CatalogImportApi {
     };
   }
 
+  preparePullRequest(): {
```
does this not want to be async? for the future

Fox32 replied:
😆 Yeah, thought that too. The return values are used as the initial state. The current usage requires it to be sync, but I think I could rewrite that a bit.

Fox32 replied:
Here we go, now it's sync

freben reviewed (approved):
yep looks good, one question

freben commented on plugins/catalog-import/src/components/ImportInfoCard/ImportInfoCard.test.tsx:
```diff
@@ -0,0 +1,101 @@
+      new ConfigReader({ integrations: {} }),
+    ).with(catalogImportApiRef, catalogImportApi);
+
+    await act(async () => {
```
hint: `await renderInTestApp` is a superb `act` + `wrapInTestApp` + `render`, and can probably be used pretty much always, especially instead of the wrap+render duo. It also returns the outcome, so you could place the `expect` on the outside

Fox32 replied:
Just copied that code around, but that pattern is indeed a bit strange. Went through all files to fix it!

freben commented on plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.test.tsx:
```diff
@@ -14,23 +14,25 @@
 import { act, render, screen } from '@testing-library/react';
 import userEvent from '@testing-library/user-event';
 import React from 'react';
+import { errorApiRef } from '../../../../../packages/core-plugin-api/src';
```
oh no!

Fox32 replied:
Lint would be pretty cool here 😆 Sometimes vscode likes to produce these direct imports…

freben replied:
https://github.com/backstage/backstage/blob/fa1339e227a558f126879f0693101d68c3e5a2c2/packages/cli/config/eslint.js?plain=1#L86

https://github.com/backstage/backstage/blob/fa1339e227a558f126879f0693101d68c3e5a2c2/packages/cli/config/eslint.backend.js?plain=1#L65

but they don't capture all cases :/ improvements are welcome

freben reviewed (changes requested):
We're close now!

freben reviewed (approved):
Alright, just the CI error left then

Fox32 commented:
> Alright, just the CI error left then

Interesting, I removed an `await` where vscode told me that it doesn't make sense, but it actually returns a promise.

![image](https://user-images.githubusercontent.com/648527/133748855-9cf25b95-1ca6-4e40-8bac-bf41ed01c521.png)

Anyway, fixed now, let's wait for the build and then it's ready
